presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java

```diff
@@ -0,0 +1,328 @@
+    }
+
+    @Override
+    public Expression rewriteWhenClause(WhenClause node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteCoalesceExpression(CoalesceExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteInListExpression(InListExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        FunctionCall rewrittenFunctionCall = (FunctionCall) updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+
+        if (rewrittenFunctionCall != node) {
+            Signature signature = analysis.getFunctionSignature(node);
+            analysis.addFunctionSignature(rewrittenFunctionCall, signature);
+
+            return rewrittenFunctionCall;
+        }
+
+        return node;
+    }
+
+    @Override
+    public Expression rewriteLambdaExpression(LambdaExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteLikePredicate(LikePredicate node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteInPredicate(InPredicate node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteArrayConstructor(ArrayConstructor node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteSubscriptExpression(SubscriptExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteCast(Cast node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
+    }
+
+    @Override
+    public Expression rewriteTryExpression(TryExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return updateTypeIfRewritten(treeRewriter.defaultRewrite(node, context), node);
```

Comment:
this bunch of methods looks like a hack and it is error prone. Also I dislike that you modify `Analysis` object in the planning phase (call this rewrite in `QueryPlanner` etc).

Would it be possible to rewrite grouping funciton as `StatementRewrite`. This happens before semantic analysis, so you would not touch analysis at all with such approach.